### DISCUSSION TITLE
Vamos falar sobre Wlang!

### DISCUSSION THREAD

**Discussion by HashLDash** [HashLDash/Photon#3] [General]
Aqui é o lugar para falar sobre a linguagem, o que vocês acham, o que seria legal ter, etc. As discussões relevantes para o desenvolvimento da linguagem vão virar issues, que guiarão o desenvolvimento.

**Comment by ntaraujo** [comment]
Eae will! Daria pra importar modulos python no wlang? Se sim teria limitações tais como as do PyPy?

**Reply by HashLDash** [reply]
Olá! Sim, é possível! A limitação é a velocidade (limitada ao python) e a distribuição do código, que estará limitada a distribuição do pacote que vc usar.

**Reply by HashLDash** [reply]
Se for um módulo em python puro e que não usa introspecção, dá pra até importar ele de maneira otimizada, deixando ele muito mais rápido. Mas, por enquanto, só dá pra fazer isso com código bem simples.

**Comment by ntaraujo** [comment]
Eae will! Man tenho uma ideia de sintaxe pra lang
mas nn sei se ainda é mt cedo pra sugerir kk

**Reply by HashLDash** [reply]
Opa, pode falar!

**Reply by ntaraujo** [reply]
Ent, muitas vezes precisamos utilizar as variáveis não só como recipientes para valores, mas algo com controle maior. Observar quando o valor é requisitado, quando muda, sincronizar duas variáveis, etc. Para isso o python tem o `property`, mas sua utilização em larga escala pode ser cansativa.
A gnt sabe q o kivy achou uma ótima solução pra isso. É só colocar
```py
text: "Variável que eu quero mostrar: " + str(self.var)
```
no seu `.kv` e pronto. O `Builder` se encarrega de criar uma `StringProperty` e dar `bind` nas variáveis citadas. Toda vez q `self.var` mudar o `self.text` muda tbm.

A sugestão é roubar parcialmente essa ideia kkk

Um código no python q se pareceria com isso:

```py
class Foo:
    def __init__(self):
        self._x = None

    def getx(self):
        return self._x

    def setx(self, value):
        if value == 'acceptable':
            self._x = value

    x = property(getx, setx)
```
poderia usar a sintaxe parecida com a dos `.kv` e virar simplesmente:
```py
_x = None
x: _x

on x as value:  # def on_x(value):
    if value == 'acceptable':
        _x = value
```
Nesse caso `getx` é substituido pelos dois pontos, e o `setx` por uma palavra reservada `on` para significar que você quer rodar um bloco de código quando um valor é atribuido à variável especificada.

Um `bind` automático no interpretador da linguagem daria superpoderes se comparado à uma simples `property` do python.
Pois se o getter é muito complexo mas os valores não mudam, os recursos da máquina são desperdiçados. A função vai ser chamada em cada requisição do valor.
Mas se o interpretador conseguir começar a observar uma variável, quando o código pede, o getter só vai ser chamado quando realmente precisa.

Em python:
```py
from some_hard_module import some_hard_method
from time import perf_counter


class Foo:
    def __init__(self):
        self._x = None

    def getx(self):
        return some_hard_method(self._x)

    def setx(self, value):
        if value == 'acceptable':
            self._x = value

    x = property(getx, setx)


foo = Foo()

tic = perf_counter()
for _ in range(1000):
    _ = foo.x
tac = perf_counter()
print(tac - tic, 'in this iteration')
# slow
```

Em wlang (se pá :P):
```py
from some_hard_module import some_hard_method
from time import perf_counter


_x = None
x: some_hard_method(_x)

on x as value:
    if value == 'acceptable':
        _x = value

tic = perf_counter()
for _ in 0..1000:
    _ = x
tac = perf_counter()
print(tac - tic, 'in this iteration')
# fast
```

**Reply by HashLDash** [reply]
Isso é basicamente o Observer Pattern. Eu acho que a notação property(getx, setx) é bem compacta, intuitiva e não depende de nenhuma adição de sintaxe. Com toda certeza, no wlang isso será muito mais rápido.

**Reply by ntaraujo** [reply]
pensei q seria legal pois foi uma necessidade q o kivy encontrou, e teve q resolver o problema em uma sublinguagem lado a lado com o python, já q o cobrinha era limitado para esse aspecto.
qnd to usando o kivy, eu raramente acho fácil e intuitivo escrever uma `AliasProperty` na minha classe, junto com um getter e um setter... Polui muito visualmente.
ent tento usar o q é oferecido nos `.kv`, mas é claro q existem GRANDES limitações de lógica ali dentro, sendo necessário tbm referenciar um método ou variável em um arquivo, trazer pro outro e vice-versa

mas vc q sabe man, se acha q não se encaixa ou q é difícil de aplicar (não sei quais limitações o projeto tem por causa da base dele) segue aí 😄

**Reply by HashLDash** [reply]
Isso será feito em algum momento. Interfaces gráficas usam bastante esse tipo de conceito. A minha dúvida é em relação a criar uma nova sintaxe para isso. Cada framework implementa de um jeito. O kivy usa essas properties, o flutter usa o setState. Ainda não sei a melhor maneira para implementar isso no Wlang/Wui. Isso pois precisa ser um método compatível com todas as frameworks/linguagens, e não afetar muito o desempenho.

**Reply by HashLDash** [reply]
Do jeito que o Wui está, não é necessária nenhuma adição de propriedade. Mas a implementação desse pattern pode deixar os apps mais eficientes energeticamente.

**Reply by ntaraujo** [reply]
Wui é o novo nome da lang? 😮

**Reply by HashLDash** [reply]
Não hehe, é o nome da biblioteca gráfica Wlang User Interface (Wui)

**Reply by ntaraujo** [reply]
GOD eu n tava sabendo! Vc postou no canal?

**Reply by HashLDash** [reply]
Falei na live de como as linguagens funcionam [aqui](https://www.youtube.com/watch?v=NpNBkjLXLUw).

**Reply by HashLDash** [reply]
Estou colocando aos poucos os arquivos aqui. Vou fazer um repositorio separado para o Wui. Ou Pui agora... sei lá o nome :P

**Comment by robertoweller** [comment]
Opa Will, acha interessante em deixar o README.md com duas línguas?  Uma em português e outra em inglês.

**Reply by robertoweller** [reply]
Seria uma tentativa de atrair mais Brasileiros e não deixaria de atender que existe outros países.

**Reply by HashLDash** [reply]
Boa! Já dei merge!

**Comment by HashLDash** [comment]
Com a linguagem funcionando, estou colocando os arquivos aqui aos poucos. Isso porque pretendo mudar algumas coisas para facilitar e melhorar o código e seu entendimento. Para ficar amigável às contribuições hehe.

**Comment by haru02w** [comment]
Desculpe-me por uma pergunta boba como esta mas...  Não consigo conter minha curiosidade.

Porque essa linguagem é tão rápida? O proprio Linux usa C, e todos à recomendam por ser mais baixo nível e, consequentemente veloz.

O que essa linguagem faz (ou o que ela deixa de fazer) em relação a tantas linguagens conceituadas e utilizadas amplamente por grandes e inumeras empresas para existir tal discrepancia de performance?

**Reply by ntaraujo** [reply]
Primeiro, parabéns pela qualidade de escrita.
Acho que o Will vai complementar melhor, mas é tudo uma questão sobre otimização.
Mesmo com o C, por não ser código de máquina, mas sim uma sintaxe com abstrações, sempre existem aspectos do código a serem otimizados.
Pedaços repetitivos ou desnecessários, que funcionam, mas que uma compreensão melhor sobre como o problema pode ser resolvido resulta em um código mais rápido.
Quando falamos de compilar o C, parece que não há um meio termo, é de tão baixo nível que passa direto pelo compilador pra linguagem de máquina. Mas na verdade ele É transpilado para Assembly durante esse processo.
A maneira que qualquer compilador aplica otimização é essa: entende a abstração da linguagem e linca ideias (necessidades) com as instruções assembly que resolvem esse problema.
Pelo que eu entendo, o Photon pode ser mais rápido que o C por isso. Ele entende uma ideia abstrata e traduz esse entendimento para outra linguagem (isso é importante, traduz a ideia e não a implementação).
Essa explicação também se aplica ao Haxe. Como a gente viu na live ele demorou séculos de análise mas produziu um código em C++ mais rápido que o Photon, o C, o C++ puro e etc.

**Reply by haru02w** [reply]
Então, rudemente, posso dizer que o Photon é um tradutor e otimizador de abstrações de outras linguagens utilizando uma sintaxe própria parecida com python?

Exemplo: Escrevo um codigo em photon, ele reescreve de forma otimizada em outra linguagem, compila (caso necessario) e executa. Certo? 

O Haxe faz o mesmo processo?
(ainda estou no ensino médio e é meio difícil entender conceitos tão abstratos)

**Reply by ntaraujo** [reply]
basicamente, sim (também no médio amigo)

**Reply by robertoweller** [reply]
@Haru02w pelo que já entendi, python quando você declara uma variável é criado um dicionário, dentro desse dicionário tem todos os tipos que ela pode ser, tem até um Len(), em caso de listas dentro dele, Photon passa a faca, e declara só sendo um tipo. Posso ter falado alguma bobagem, mas tem mais coisas que ele faz.

Uma parte boa parte dele é em python, deve ter alguma coisa em C, eu parei de olhar a atualização do Photon, mas recomendo da um olhada no código dele, mesmo que não entenda algo é bom para começar entender.

**Reply by Huyston** [reply]
É basicamente isso @Haru02w 
Tem algumas outras coisas no meio. Por exemplo, vc pode criar executaveis para diferentes plataformas de maneira automatizada. E também abstrai as apis especifixas da plataforma.
